File: web/types/events.ts
export type EventType =
    | 'agent_start'
    | 'check_if_news_text_start'
    | 'check_if_news_text_end'
    | 'extract_basic_metadata_start'
    | 'extract_basic_metadata_end'
    | 'extract_knowledge_start'
    | 'extract_knowledge_end'
    | 'retrieve_knowledge_start'
    | 'retrieve_knowledge_end'
    | 'extract_check_point_start'
    | 'extract_check_point_end'
    | 'search_agent_start'
    | 'evaluate_current_status_start'
    | 'evaluate_current_status_end'
    | 'generate_answer_start'
    | 'generate_answer_end'
    | 'evaluate_search_result_start'
    | 'evaluate_search_result_end'
    | 'llm_decision'
    | 'write_fact_check_report_start'
    | 'write_fact_check_report_end'
    | 'tool_start'
    | 'tool_end'
    | 'task_complete'
    | 'task_interrupted'
    | 'error'
    | 'stream_closed';
export const eventTypes: EventType[] = [
    'agent_start',
    'check_if_news_text_start', 'check_if_news_text_end',
    'extract_check_point_start', 'extract_check_point_end',
    'extract_basic_metadata_start', 'extract_basic_metadata_end',
    'extract_knowledge_start', 'extract_knowledge_end',
    'retrieve_knowledge_start', 'retrieve_knowledge_end',
    'search_agent_start', 'evaluate_current_status_start', 'evaluate_current_status_end',
    'tool_start', 'tool_end',
    'generate_answer_start', 'generate_answer_end',
    'evaluate_search_result_start', 'evaluate_search_result_end',
    'write_fact_check_report_start', 'write_fact_check_report_end',
    'llm_decision', 'task_complete', 'task_interrupted',
    'error',
];


export interface Event<T=any> {
    event: EventType;
    data?: T;
}

// IsNewsText interface
export interface IsNewsText {
    result: boolean;
    reason: string;
}

// Main Agent Events
export interface RetrievalStep {
    id: string;
    purpose: string;
    expected_source: string;
    result?: RetrievalResult;
    verification?: RetrievalResultVerification;
}

export interface CheckPoint {
    id: string;
    content: string;
    is_verification_point: boolean;
    importance?: string;
    retrieval_step?: RetrievalStep[];
}

// Metadata Extractor Events
export interface BasicMetadata {
    news_type: string;
    who?: string[];
    when?: string[];
    where?: string[];
    what?: string[];
    why?: string[];
    how?: string[];
}
export interface Knowledge {
    term: string;
    category: string;
    description?: string;
    source?: string;
}

// Search Agent Events
export interface SearchAgentStartData {
    content: string;
    purpose: string;
    expected_source: string;
}

export interface Status {
    evaluation: string;
    missing_information: string;
    new_evidence: Evidence[];
    next_step: string;
}

export interface Evidence {
    content: string;
    source: Record<string, string>;
    reasoning: string;
    relationship: "support" | "contradict";
}

export interface SearchResult {
    summary: string;
    conclusion: string;
}
export interface RetrievalResult extends SearchResult {
    check_point_id: string;
    retrieval_step_id: string;
    evidences: Evidence[];
}
export interface RetrievalResultVerification {
    reasoning: string;
    verified?: boolean;
    updated_purpose?: string;
    updated_expected_source?: string;
}

export interface ToolStartData {
    tool_name: string;
    input_str: string;
}

export interface ToolEndData {
    tool_name: string;
    output_str: string;
}

export interface LLMDecisionData {
    decision: string;
    reason: string | null;
}

export interface FactCheckResultData {
    report: string;
    verdict: "true" | "mostly-true" | "mostly-false" | "false" | "no-enough-evidence";
}

export interface TaskCompleteData {
    message: string;
    result: any; // This could be further typed if the structure is consistent
}

export interface TaskInterruptedData {
    message: string;
}

export interface ErrorData {
    message: string;
}

File: puzzle/api/events.py
"""
Agent SSE Event Models
"""

import re
from pydantic import BaseModel
from agents.main.states import CheckPoint, RetrievalResultVerification, IsNewsText, Result
from agents.metadata_extractor.states import BasicMetadata, Knowledge
from agents.searcher.states import Status, SearchResult

from typing import Optional, Any, List


def convert_name_to_event(name: str) -> str:
    """
    example:
    OnExtractCheckPointStart -> on_extract_check_point_start
    """
    # 将驼峰式命名转换为下划线式命名
    s1 = re.sub("(.)([A-Z][a-z]+)", r"\1_\2", name)
    s2 = re.sub("([a-z0-9])([A-Z])", r"\1_\2", s1)
    # 转换为小写并返回
    return s2.lower()


class BaseEvent(BaseModel):
    event: Optional[str] = None
    data: Any

    def __init__(self, **data):
        if "event" not in data:
            data["event"] = convert_name_to_event(self.__class__.__name__)
        super().__init__(**data)


class OnAgentStart(BaseEvent):
    data: None = None


# check_if_news_text
class CheckIfNewsTextStart(BaseEvent):
    data: None = None


class CheckIfNewsTextEnd(BaseEvent):
    data: IsNewsText


# invoke_metadata_extract_agent
class ExtractBasicMetadataStart(BaseEvent):
    data: None = None


class ExtractBasicMetadataEnd(BaseEvent):
    data: BasicMetadata


class ExtractKnowledgeStart(BaseEvent):
    data: None = None


class ExtractKnowledgeEnd(BaseEvent):
    data: List[Knowledge]


class RetrieveKnowledgeStart(BaseEvent):
    data: None = None


class RetrieveKnowledgeEnd(BaseEvent):
    data: Knowledge


class ExtractCheckPointStart(BaseEvent):
    data: None = None


class ExtractCheckPointEnd(BaseEvent):
    data: List[CheckPoint]


# invoke_search_agent
class SearchAgentInput(BaseModel):
    content: str
    purpose: str
    expected_source: str


class SearchAgentStart(BaseEvent):
    data: SearchAgentInput


class EvaluateCurrentStatusStart(BaseEvent):
    data: None = None


class EvaluateCurrentStatusEnd(BaseEvent):
    data: Status


class GenerateAnswerStart(BaseEvent):
    data: None = None


class GenerateAnswerEnd(BaseEvent):
    data: SearchResult


class EvaluateSearchResultStart(BaseEvent):
    data: None = None


class EvaluateSearchResultEnd(BaseEvent):
    data: RetrievalResultVerification


class LLMDecisionData(BaseModel):
    decision: str


class LLMDecision(BaseEvent):
    data: LLMDecisionData


class WriteFactCheckReportStart(BaseEvent):
    data: None = None


class FactCheckResultData(BaseModel):
    report: str
    verdict: str


class WriteFactCheckReportEnd(BaseEvent):
    data: FactCheckResultData


class ToolStartData(BaseModel):
    tool_name: str
    input_str: str


class ToolStart(BaseEvent):
    data: ToolStartData


class ToolEndData(BaseModel):
    tool_name: str
    output_str: str


class ToolEnd(BaseEvent):
    data: ToolEndData


class TaskComplete(BaseEvent):
    data: None = None


class InterruptData(BaseModel):
    message: str


class TaskInterrupted(BaseEvent):
    data: InterruptData


class ErrorData(BaseModel):
    message: str


class Error(BaseEvent):
    data: ErrorData
